### ISSUE TITLE
Text input keys are partially wrong with an exotic keyboard layout

### ISSUE THREAD

**Issue by plule** [DFHack/dfhack#3525]
Using the bépo keyboard layout (french ergonomic, similar to dvorak in goal), some of the keys are wrongly mapped when typing in DFHack text boxes such as `buildingplan` filter or `gui/launcher`, and I can't type numbers. The weirdest part is that many keys are actually right. This is new to SDL2.

Expected layout:

![kbd-expected](https://github.com/DFHack/dfhack/assets/630159/0ef4a0c8-ac5d-40eb-9e5f-fe6158152e5a)

Actual layout:

![kbd-actual](https://github.com/DFHack/dfhack/assets/630159/59a6df39-4ea0-4b7d-a0aa-e8b31d9ef4ab)

The Dwarf Fortress text inputs behave as expected (with some variations on the special characters)

**Comment by myk002** [comment]
Could you run this script:
```lua
local gui = require('gui')
local widgets = require('gui.widgets')

KeyScreen = defclass(KeyScreen, gui.ZScreenModal)
KeyScreen.ATTRS {
    focus_path='keydump',
}

function KeyScreen:onInput(keys)
    print('keys:')
    printall(keys)
    print('text:')
    print(df.global.enabler.last_text_input[0])
    KeyScreen.super.onInput(self, keys)
end

KeyScreen():show()
```

and report the output for a few keys that behave correctly and a few keys that don't (including number keys)?

**Comment by plule** [comment]
Oops, the numbers are actually OK in `gui/control-panel`, not in the building plan filter - I assume it's normal then, no need to type numbers there.

# Bad

"R" physical key, expected: "O"

```
keys:
STRING_A111              = true
HOTKEY_BUILDING_MACHINE_ROLLERS  = true
_STRING                  = 114
HOTKEY_BUILDING_WORKSHOP_CRAFTSMAN       = true
ARENA_FLUIDS_REMOVE      = true
DESIGNATE_RAMP           = true
STRING_A114              = true
HOTKEY_BUILDING_CHAIR    = true
HOTKEY_BUILDING_CAGES_CHAINS     = true
HOTKEY_BUILDING_WEAPONRACK       = true
ARENA_CREATURE_RESET     = true
CUSTOM_R                 = true
DESIGNATE_TRAFFIC_RESTRICTED     = true
HOTKEY_BUILDING_DOOR     = true
D_TOGGLE_RAMP_INDICATORS         = true
HOTKEY_BUILDING_CONSTRUCTION_RAMP        = true
text:
111
```

"F" physical key, expected "E"

```
keys:
STRING_A102              = true
HOTKEY_BUILDING_WORKSHOP_FARMER  = true
CUSTOM_F                 = true
D_TOGGLE_FLUID_NUMBERS   = true
ARENA_FLUIDS             = true
_STRING                  = 102
STRING_A101              = true
HOTKEY_BUILDING_FURNITURE        = true
HOTKEY_BUILDING_FLOODGATE        = true
HOTKEY_BUILDING_FARMING  = true
HOTKEY_BUILDING_FURNACE_WOOD     = true
HOTKEY_BUILDING_CONSTRUCTION_FLOOR       = true
DESIGNATE_FORTIFY        = true
DESIGNATE_UNCLAIM        = true
text:
101
```

"H" physical key, expected "C"

```
keys:
HOTKEY_BUILDING_WORKSHOP_MECHANIC        = true
HOTKEY_BUILDING_HIVE     = true
CUSTOM_H                 = true
DESIGNATE_TRAFFIC_HIGH   = true
STRING_A099              = true
_STRING                  = 104
HOTKEY_BUILDING_BOX      = true
DESIGNATE_HIDE           = true
HOTKEY_BUILDING_CHAIN    = true
HOTKEY_BUILDING_HATCH    = true
STRING_A104              = true
D_HOT_KEYS               = true
HOTKEY_BUILDING_MACHINE_WATER_WHEEL      = true
text:
99
```

# Good

"E" physical key, types as expected "P"

```
keys:
CUSTOM_E                 = true
STRING_A101              = true
STRING_A112              = true
CURSOR_UP_Z              = true
_STRING                  = 112
text:
112
```

"K" physical key, types as expected "S"
```
keys:
DESIGNATE_MARKER_TO_STANDARD     = true
CUSTOM_K                 = true
D_STOCKS                 = true
STRING_A115              = true
HOTKEY_BUILDING_WORKSHOP_CLOTHES         = true
_STRING                  = 115
HOTKEY_BUILDING_CONSTRUCTION_TRACK       = true
STRING_A107              = true
HOTKEY_BUILDING_WORKSHOP_KITCHEN         = true
ARENA_CREATE_CREATURE    = true
HOTKEY_BUILDING_FURNACE_KILN     = true
text:
115
```

"L" physical key, types as expected "R"
```
keys:
D_DESIGNATE_CHOP         = true
STRING_A108              = true
HOTKEY_BUILDING_SLAB     = true
HOTKEY_BUILDING_TRAP_LEVER       = true
HOTKEY_BUILDING_FURNACE_SMELTER  = true
HOTKEY_BUILDING_WORKSHOP_LEATHER         = true
DESIGNATE_TRAFFIC_LOW    = true
HOTKEY_BUILDING_WORKSHOPS_CLOTHING_LEATHER       = true
STRING_A114              = true
DESIGNATE_CHOP           = true
CUSTOM_L                 = true
_STRING                  = 114
HOTKEY_BUILDING_WORKSHOP_STILL   = true
HOTKEY_BUILDING_CONSTRUCTION_WALL        = true
text:
114
```

1: physical key maj+1 (like azerty), types as expected 1:

```
keys:
KEYBOARD_CURSOR_DOWNLEFT_FAST    = true
STRING_A049              = true
_STRING                  = 49
text:
49
```

**Comment by myk002** [comment]
Ok, so the issue looks like there are multiple `STRING_A???` values, and `_STRING` is picking up the wrong one in the "bad" cases.

This may actually get fixed by #3527
Could you try pulling the build from there and seeing if that fixes things?

**Comment by plule** [comment]
@myk002 the build 44f11bb that just finished fixes it, thanks a lot!
